### PR TITLE
Edition fixups: remove `extern crate`, add idioms lint

### PR DESCRIPTION
**Pull request by tarcieri**
Rust editions 2018+ do not require `extern crate` except for linking `alloc` and `std`.

### CODE REVIEW

File: src/batch.rs [hunk 1]
```diff
@@ -9,12 +9,7 @@
 
 //! Batch signature verification.
 
-#[cfg(feature = "alloc")]
-extern crate alloc;
-#[cfg(feature = "alloc")]
 use alloc::vec::Vec;
-#[cfg(all(not(feature = "alloc"), feature = "std"))]
-use std::vec::Vec;
```
Comment (tarcieri) [inline src/batch.rs:17]: Changed these to use `alloc` unconditionally.

`alloc` is already a hard requirement of batch verification.

File: Cargo.toml [hunk 2]
```diff
@@ -48,13 +48,13 @@ required-features = ["batch"]
...
-std = ["curve25519-dalek/std", "ed25519/std", "serde_crate/std", "sha2/std", "rand/std"]
+std = ["alloc", "curve25519-dalek/std", "ed25519/std", "serde_crate/std", "sha2/std", "rand/std"]
 alloc = ["curve25519-dalek/alloc", "rand/alloc", "zeroize/alloc"]
 nightly = ["curve25519-dalek/nightly"]
 serde = ["serde_crate", "serde_bytes", "ed25519/serde"]
-batch = ["merlin", "rand/std"]
+batch = ["alloc", "merlin", "rand/std"]
 # This feature enables deterministic batch verification.
-batch_deterministic = ["merlin", "rand", "rand_core"]
+batch_deterministic = ["alloc", "merlin", "rand", "rand_core"]
```
Comment (tarcieri) [inline Cargo.toml:57]: Batch verification has a hard requirement on `alloc`

**Comment by tarcieri** [comment]
Hmm, weird, running into this (seemingly unrelated) error:

https://github.com/dalek-cryptography/ed25519-dalek/actions/runs/3633662766/jobs/6130930018#step:4:140

```
error: simd_backend selected without target_feature=+avx2 or +avx512ifma
  --> /home/runner/.cargo/git/checkouts/curve25519-dalek-3aa0eb2fdf9e336b/e01bb1b/src/backend/vector/mod.rs:15:1
   |
15 | compile_error!("simd_backend selected without target_feature=+avx2 or +avx512ifma");
```

File: src/lib.rs [hunk 3]
```diff
@@ -232,40 +205,34 @@
...
-
-#[cfg(all(feature = "alloc", not(feature = "std")))]
-extern crate alloc;
-extern crate curve25519_dalek;
-#[cfg(all(any(feature = "batch", feature = "batch_deterministic"), any(feature = "std", feature = "alloc")))]
-extern crate merlin;
-#[cfg(any(feature = "batch", feature = "std", feature = "alloc", test))]
-extern crate rand;
 #[cfg(feature = "serde")]
 extern crate serde_crate as serde;
```
Comment (tarcieri) [inline src/lib.rs:220]: Can't get rid of this yet, unfortunately, but we can probably get rid of it by switching to namespaced features

File: tests/ed25519.rs [hunk 4]
```diff
@@ -32,25 +23,28 @@ mod vectors {
...
+        // TestGolden
         let mut line: String;
         let mut lineno: usize = 0;
 
         let f = File::open("TESTVECTORS");
         if f.is_err() {
-            println!("This test is only available when the code has been cloned \
+            println!(
+                "This test is only available when the code has been cloned \
                       from the git repository, since the TESTVECTORS file is large \
```
Comment (rozbb) [inline tests/ed25519.rs:45]: Weirdly indented

File: tests/ed25519.rs [hunk 5]
```diff
@@ -112,11 +109,19 @@ mod vectors {
...
 
-        assert!(sig1 == sig2,
-                "Original signature from test vectors doesn't equal signature produced:\
-                \noriginal:\n{:?}\nproduced:\n{:?}", sig1, sig2);
-        assert!(keypair.verify_prehashed(prehash_for_verifying, None, &sig2).is_ok(),
-                "Could not verify ed25519ph signature!");
+        assert!(
+            sig1 == sig2,
+            "Original signature from test vectors doesn't equal signature produced:\
+                \noriginal:\n{:?}\nproduced:\n{:?}",
```
Comment (rozbb) [inline tests/ed25519.rs:115]: Weirdly indented. Is this a norm I don't know about?

File: tests/ed25519.rs [hunk 6]
```diff
@@ -197,38 +202,45 @@ mod integrations {
...
         let good_sig: Signature;
-        let bad_sig:  Signature;
+        let bad_sig: Signature;
 
         let good: &[u8] = b"test message";
-        let bad:  &[u8] = b"wrong message";
+        let bad: &[u8] = b"wrong message";
 
-        let mut csprng = OsRng{};
+        let mut csprng = OsRng {};
```
Comment (rozbb) [inline tests/ed25519.rs:243]: Can remove braces entirely.

File: tests/ed25519.rs [hunk 7]
```diff
@@ -268,7 +296,7 @@ mod integrations {
             b"Fuck dumbin' it down, spit ice, skip jewellery: Molotov cocktails on me like accessories.",
             b"Hey, I never cared about your bucks, so if I run up with a mask on, probably got a gas can too.",
             b"And I'm not here to fill 'er up. Nope, we came to riot, here to incite, we don't want any of your stuff.", ];
-        let mut csprng = OsRng{};
+        let mut csprng = OsRng {};
```
Comment (rozbb) [inline tests/ed25519.rs:299]: Can remove braces entirely

File: .github/workflows/rust.yml [hunk 8]
```diff
@@ -32,13 +32,18 @@ jobs:
       - run: cargo test --target ${{ matrix.target }} --features batch_deterministic
       - run: cargo test --target ${{ matrix.target }} --features serde
 
-  test-simd:
+  build-simd:
```
Comment (rozbb) [inline .github/workflows/rust.yml:35]: I think this part was covered in the #232. Can you rebase?

**Review by rozbb — COMMENTED** [review]
Looks great. We should do a proper clippy pass some time before 2.0 if we can

**Comment by tarcieri** [comment]
Hmm weird, it seems like rustfmt or something got applied against some of the files I touched (I do have my editor configured to run it on save, since every other project I work within uses it)

File: tests/ed25519.rs (hunk 5 above)
Comment (tarcieri) [reply tests/ed25519.rs:115]: Nah, just the result of rustfmt being run across the code

**Pull request merged**